### PR TITLE
feat: add tally/stateful-root-runtime rule

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

- Add `tally/stateful-root-runtime` rule — detects final stages running as root when the Dockerfile signals mutable/persistent state (VOLUME, data directories). More targeted than `hadolint/DL3002`: only fires when root intersects with state that may outlive the container.
- Extend facts framework with `StageFacts.EffectiveUser`, `UserCommands`, `Volumes`, `FinalWorkdir`, `HasPrivilegeDropEntrypoint`, and exported `IsRootUser` helper — foundation for 7+ planned USER-related rules from the design doc.
- Refactor `hadolint/DL3002` to consume shared facts instead of walking instructions directly, removing duplicated `isRootUser` logic.

### Detection signals

| Signal | Confidence |
|--------|-----------|
| `VOLUME` instructions | High |
| `WORKDIR` to `/data`, `/var/lib/*`, `/var/log/*`, `/var/cache/*`, `/var/run/*`, `/srv` | Medium |
| `COPY`/`ADD` destination to state paths | Medium |
| `RUN mkdir` of state directories | Medium |

### Suppression

- Privilege-drop tools in ENTRYPOINT/CMD (`gosu`, `su-exec`, `suexec`, `setpriv`)
- Known non-root base images (distroless `:nonroot`, chainguard/cgr.dev)
- Explicit non-root `USER` instruction
- Inherited non-root USER from parent stage

### Follow-up

- #382 tracks extending privilege-drop detection to inspect entrypoint script content (inline heredocs and build-context files) via a unified `StageFacts.FileContent(path)` primitive

## Test plan

- [x] 15 new facts tests (IsRootUser, user/volume tracking, privilege-drop, multi-stage isolation)
- [x] 32 rule unit tests (violations, clean cases, suppression, multi-stage, edge cases)
- [x] Integration test with snapshot
- [x] DL3002 tests pass unchanged after refactor
- [x] `go test ./...` passes
- [x] `make lint` clean (0 issues)
- [x] `make cpd` clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added security rule tally/stateful-root-runtime to warn when final runtime stages run as root alongside stateful operations (volumes, persistent directories).

* **Documentation**
  * Updated README/docs to reflect 41 tally rules and added documentation, examples, and configuration for the new rule.

* **Tests**
  * Added and updated unit/integration tests, snapshots, and expected diagnostics covering root detection, privilege-drop behavior, multi-stage scenarios, and state-path detection.

### CODE REVIEW

File: internal/rules/tally/stateful_root_runtime.go [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+		return ""
+	}
+
+	info := model.StageInfo(stageIdx)
+	if info == nil || info.BaseImage == nil || !info.BaseImage.IsStageRef || info.BaseImage.StageIndex < 0 {
+		return ""
+	}
+
+	if parentFacts := fileFacts.Stage(info.BaseImage.StageIndex); parentFacts != nil {
+		return parentFacts.FinalWorkdir
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/stateful_root_runtime.go]: ✅ Addressed in aa4a2ca. `inheritedParentWorkdir` now walks the full stage-ref chain with cycle guard, so `base (WORKDIR /var) -> mid (no WORKDIR) -> final` correctly inherits `/var`. Added regression test for transitive chain. Thanks @chatgpt-codex-connector\!
